File: apps/ios/Shared/MissionCatalog.swift
import Foundation

/// Missões diárias locais (offline) — espelham a API.
struct LocalMission: Codable, Identifiable, Equatable {
  var id: String
  var kind: String
  var title: String
  var description: String
  var target: Int
  var progress: Int
  var rewardEnergy: Int
  var rewardAffection: Int
  var claimed: Bool

  var complete: Bool { progress >= target }
}

enum MissionCatalog {
  private static let storeKey = "companion.missions.v1"

  private static let rotation: [[(kind: String, title: String, description: String, target: Int, e: Int, a: Int)]] = [
    [
      ("POKE_COUNT", "Cutucadas", "Cutuca o dino 3 vezes", 3, 6, 4),
      ("FEED_COUNT", "Lanche", "Alimente 2 vezes", 2, 10, 3),
      ("PLAY_COUNT", "Brincadeira", "Brinque 1 vez", 1, 8, 6),
    ],
    [
      ("CHAT_COUNT", "Conversa", "Mande 1 mensagem", 1, 5, 8),
      ("TEASE_COUNT", "Piadinha", "Mande 1 piada", 1, 7, 7),
      ("OPEN_APP", "Visita", "Abra o app 2 vezes hoje", 2, 4, 5),
    ],
    [
      ("POKE_COUNT", "Carinho", "Cutuca 5 vezes", 5, 8, 5),
      ("FEED_COUNT", "Banquete", "Alimente 3 vezes", 3, 12, 4),
      ("TEASE_COUNT", "Zoeira", "2 piadas no dia", 2, 9, 9),
    ],
  ]

  static func dayKey(_ date: Date = Date()) -> String {
    let f = ISO8601DateFormatter()
    f.formatOptions = [.withFullDate]
    return f.string(from: date)
  }

  static func ensureToday() -> [LocalMission] {
    let key = dayKey()
    if let existing = load(), existing.dayKey == key {
      return existing.missions
    }
    let n = key.split(separator: "-").compactMap { Int($0) }.reduce(0, +)
    let defs = rotation[n % rotation.count]
    let missions = defs.map { d in
      LocalMission(
        id: "local-\(key)-\(d.kind)",
        kind: d.kind,
        title: d.title,
        description: d.description,
        target: d.target,
        progress: 0,
        rewardEnergy: d.e,
        rewardAffection: d.a,
        claimed: false
      )
    }
    save(dayKey: key, missions: missions)
    return missions
  }

  static func bump(kind: String, amount: Int = 1) -> [LocalMission] {
    var missions = ensureToday()
    guard let idx = missions.firstIndex(where: { $0.kind == kind }) else { return missions }
    guard !missions[idx].claimed else { return missions }
    missions[idx].progress = min(missions[idx].target, missions[idx].progress + amount)
    save(dayKey: dayKey(), missions: missions)
    return missions
  }

  static func kindFromInteraction(_ type: String) -> String? {
    switch type.uppercased() {
    case "POKE": return "POKE_COUNT"
    case "FEED": return "FEED_COUNT"
    case "PLAY": return "PLAY_COUNT"
    case "CHAT": return "CHAT_COUNT"
    case "TEASE": return "TEASE_COUNT"
    default: return nil
    }
  }

  @discardableResult
  static func claim(_ id: String) -> (missions: [LocalMission], rewardEnergy: Int, rewardAffection: Int)? {
    var missions = ensureToday()
    guard let idx = missions.firstIndex(where: { $0.id == id }) else { return nil }
    let m = missions[idx]
    guard !m.claimed, m.complete else { return nil }
    missions[idx].claimed = true
    save(dayKey: dayKey(), missions: missions)
    return (missions, m.rewardEnergy, m.rewardAffection)
  }

  private struct Box: Codable {
    var dayKey: String
    var missions: [LocalMission]
  }

  private static func load() -> Box? {
    guard let data = UserDefaults.standard.data(forKey: storeKey) else { return nil }
    return try? JSONDecoder().decode(Box.self, from: data)
  }

  private static func save(dayKey: String, missions: [LocalMission]) {
    let box = Box(dayKey: dayKey, missions: missions)
    if let data = try? JSONEncoder().encode(box) {
      UserDefaults.standard.set(data, forKey: storeKey)
    }
  }
}

File: apps/desktop/electron/missionCatalog.ts
/** Missões diárias — espelha apps/ios/Shared/MissionCatalog.swift e src/missions.ts */

export type MissionKind =
  | "POKE_COUNT"
  | "FEED_COUNT"
  | "PLAY_COUNT"
  | "CHAT_COUNT"
  | "TEASE_COUNT"
  | "OPEN_APP";

export type LocalMission = {
  id: string;
  kind: MissionKind;
  title: string;
  description: string;
  target: number;
  progress: number;
  rewardEnergy: number;
  rewardAffection: number;
  claimed: boolean;
};

type MissionDef = {
  kind: MissionKind;
  title: string;
  description: string;
  target: number;
  rewardEnergy: number;
  rewardAffection: number;
};

const ROTATION: MissionDef[][] = [
  [
    { kind: "POKE_COUNT", title: "Cutucadas", description: "Cutuca o dino 3 vezes", target: 3, rewardEnergy: 6, rewardAffection: 4 },
    { kind: "FEED_COUNT", title: "Lanche", description: "Alimente 2 vezes", target: 2, rewardEnergy: 10, rewardAffection: 3 },
    { kind: "PLAY_COUNT", title: "Brincadeira", description: "Brinque 1 vez", target: 1, rewardEnergy: 8, rewardAffection: 6 },
  ],
  [
    { kind: "CHAT_COUNT", title: "Conversa", description: "Mande 1 mensagem", target: 1, rewardEnergy: 5, rewardAffection: 8 },
    { kind: "TEASE_COUNT", title: "Piadinha", description: "Mande 1 piada", target: 1, rewardEnergy: 7, rewardAffection: 7 },
    { kind: "OPEN_APP", title: "Visita", description: "Abra o app 2 vezes hoje", target: 2, rewardEnergy: 4, rewardAffection: 5 },
  ],
  [
    { kind: "POKE_COUNT", title: "Carinho", description: "Cutuca 5 vezes", target: 5, rewardEnergy: 8, rewardAffection: 5 },
    { kind: "FEED_COUNT", title: "Banquete", description: "Alimente 3 vezes", target: 3, rewardEnergy: 12, rewardAffection: 4 },
    { kind: "TEASE_COUNT", title: "Zoeira", description: "2 piadas no dia", target: 2, rewardEnergy: 9, rewardAffection: 9 },
  ],
];

/** dayKey local YYYY-MM-DD (espelha iOS Calendar.current). */
export function dayKey(date = new Date(), timeZone = Intl.DateTimeFormat().resolvedOptions().timeZone): string {
  const parts = new Intl.DateTimeFormat("en-CA", {
    timeZone,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
  }).formatToParts(date);
  const y = parts.find((p) => p.type === "year")!.value;
  const m = parts.find((p) => p.type === "month")!.value;
  const d = parts.find((p) => p.type === "day")!.value;
  return `${y}-${m}-${d}`;
}

export function displayDayLabel(date = new Date()): string {
  const key = dayKey(date);
  const parts = key.split("-");
  if (parts.length !== 3) return "Hoje";
  return `Hoje · ${parts[2]}/${parts[1]}`;
}

function defsForDay(key: string): MissionDef[] {
  const n = key.split("-").reduce((a, p) => a + Number(p), 0);
  return ROTATION[n % ROTATION.length];
}

export function buildMissionsForDay(key: string): LocalMission[] {
  return defsForDay(key).map((d) => ({
    id: `local-${key}-${d.kind}`,
    kind: d.kind,
    title: d.title,
    description: d.description,
    target: d.target,
    progress: 0,
    rewardEnergy: d.rewardEnergy,
    rewardAffection: d.rewardAffection,
    claimed: false,
  }));
}

export function ensureToday(
  day: string,
  existingDay: string,
  existing: LocalMission[]
): { day: string; missions: LocalMission[] } {
  if (existingDay === day && existing.length > 0) {
    return { day, missions: existing };
  }
  return { day, missions: buildMissionsForDay(day) };
}

export function bump(
  missions: LocalMission[],
  kind: MissionKind,
  amount = 1
): LocalMission[] {
  const idx = missions.findIndex((m) => m.kind === kind);
  if (idx < 0) return missions;
  if (missions[idx].claimed) return missions;
  const next = missions.map((m, i) =>
    i === idx
      ? { ...m, progress: Math.min(m.target, m.progress + amount) }
      : m
  );
  return next;
}

export function kindFromInteraction(type: string): MissionKind | null {
  switch (String(type).toUpperCase()) {
    case "POKE":
      return "POKE_COUNT";
    case "FEED":
      return "FEED_COUNT";
    case "PLAY":
      return "PLAY_COUNT";
    case "CHAT":
      return "CHAT_COUNT";
    case "TEASE":
      return "TEASE_COUNT";
    default:
      return null;
  }
}

export function claim(
  missions: LocalMission[],
  idOrKind: string
): { missions: LocalMission[]; rewardEnergy: number; rewardAffection: number } | null {
  let idx = missions.findIndex((m) => m.id === idOrKind);
  if (idx < 0) idx = missions.findIndex((m) => m.kind === idOrKind);
  if (idx < 0) idx = missions.findIndex((m) => idOrKind.includes(m.kind));
  if (idx < 0) return null;
  const m = missions[idx];
  if (m.claimed || m.progress < m.target) return null;
  const next = missions.map((row, i) => (i === idx ? { ...row, claimed: true } : row));
  return { missions: next, rewardEnergy: m.rewardEnergy, rewardAffection: m.rewardAffection };
}
